File: 2019-2020_programmers/_15_/src/java/ShipWithinDays.java
import java.util.*;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class ShipWithinDays {
    public static void main(String[] args) {
        int[] weights =
//                {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
                {3,2,2,4,1,4};

//        int day = 5;
        int day = 3;
        System.out.println(shipWithinDays(weights, day));
    }

    public static int shipWithinDays(int[] weights, int day) {
        int avg = Arrays.stream(weights).sum() / day;
        Queue<Integer> que = (Queue<Integer>) Arrays.stream(weights).boxed().collect(Collectors.toCollection(LinkedList::new));

        Queue<Integer> clone = que;
        Supplier<LinkedList<Integer>> queProvider = () -> new LinkedList<Integer>((Collection<? extends Integer>) clone);

        int n;
        int capacity;
        do {
            que = (Queue<Integer>) queProvider.get();
            n = day;
            capacity = avg;
            /* log
             */
            System.out.println("--------- capacity : " + (capacity) + "-----------");
            while (n != 0) {
                --n;
                /* log
                 */
                System.out.println("\t▷ day : " + (day - n));
                while (true) {
                    /* log
                     */
                    System.out.println("\t- capacity: " + capacity + ", que: " + que);
                    if (que.size() == 0 || capacity - que.peek() < 0) break;
                    capacity = capacity - que.poll();
                    /* log
                     */
                    System.out.println("\t- capacity: " + capacity + ", que: " + que);
                }
                capacity = avg;
                /* log
                 */
                System.out.println();
            }
            avg += 1;
        } while (que.size() > 0);

        /* log
         */
        System.out.println("➤ min capacity: " + capacity);
        return capacity;
    }
}

File: 2019-2020_programmers/_15_/src/kotlin/shipWithinDays.kt
import java.util.function.Supplier

fun main(args: Array<String>) {
    val weights: Array<Int> = arrayOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
    val day = 5;

    println(shipWithinDaysKt(weights, day));
}

interface IQueue<T> {
    fun enqueue(item: T)
    fun poll(): T?
    fun size(): Int
    fun peek(): T?
    fun isEmpty(): Boolean
}

fun shipWithinDaysKt(weights: Array<Int>, day: Int): Int {
    var avg = weights.sum() / day

    class QueueImpl : IQueue<Int> {
        val list = mutableListOf<Int>()

        constructor(arr: Array<Int>) {
            for (i in arr) {
                list.add(i);
            }
        }

        override fun enqueue(item: Int) {
            list.add(item)
        }

        override fun poll(): Int? {
            if (!isEmpty()) {
                return list.removeAt(0)
            }; return null
        }

        override fun size(): Int {
            return list.size
        }

        override fun peek(): Int? {
            if (!list.isEmpty()) {
                return list[0]
            }; return null
        }

        override fun isEmpty(): Boolean {
            return list.size == 0
        }

        override fun toString(): String {
            return list.toString()
        }
    }

    var que: QueueImpl

    val queProvider = object : Supplier<IQueue<Int>> {
        override fun get(): IQueue<Int> {
            return QueueImpl(weights)
        }
    }

    var n: Int
    var capacity: Int
    do {
        que = queProvider.get() as QueueImpl
        n = day;
        capacity = avg;
        /* log/
         */
        println("--------- capacity :  $capacity -----------")
        while (n != 0) {
            --n;
            /* log/
             */
            println("\t▷ day : ${day - n}");
            while (true) {
                /* log /
                */
                println("\t- capacity: $capacity, que: $que");
                if (que.size() == 0 || capacity - que.peek()!! < 0) break;
                capacity -= que.poll()!!;
                /* log/
                */
                println("\t- capacity: $capacity , que: $que");
            }
            capacity = avg;
            /* log
            */
            println();
        }
        avg += 1;
    } while (que.size() > 0);

    /* log
    */
    println("➤ min capacity: $capacity");
    return capacity;
}
